tests/components/energy/test_validate.py

```diff
@@ -0,0 +1,475 @@
+    )
+
+    assert (await validate.async_validate(hass)).as_dict() == {
+        "errors": [],
+        "warnings": [],
+        "energy_sources": [],
+        "device_consumption": [
+            {
+                "errors": [
+                    {
+                        "message": "Entity sensor.unexpected_unit has an unexpected unit. Expected kWh or Wh",
+                        "link": None,
+                    }
+                ],
+                "warnings": [],
+            }
+        ],
+    }
+
+
+async def test_validation_device_consumption_recorder_not_tracked(
+    hass, mock_energy_manager, mock_is_entity_recorded
+):
+    """Test validating device based on untracked entity."""
+    mock_is_entity_recorded["sensor.not_recorded"] = False
+    await mock_energy_manager.async_update(
+        {"device_consumption": [{"stat_consumption": "sensor.not_recorded"}]}
+    )
+
+    assert (await validate.async_validate(hass)).as_dict() == {
+        "errors": [],
+        "warnings": [],
+        "energy_sources": [],
+        "device_consumption": [
+            {
+                "errors": [
+                    {
+                        "message": "Entity sensor.not_recorded needs to be tracked by the recorder",
+                        "link": "https://www.home-assistant.io/integrations/recorder#configure-filter",
+                    }
+                ],
+                "warnings": [],
+            }
+        ],
+    }
+
+
+async def test_validation_solar(hass, mock_energy_manager):
+    """Test validating missing stat for device."""
+    await mock_energy_manager.async_update(
+        {
+            "energy_sources": [
+                {"type": "solar", "stat_energy_from": "sensor.solar_production"}
+            ]
+        }
+    )
+    hass.states.async_set(
+        "sensor.solar_production", "10.10", {"unit_of_measurement": "beers"}
+    )
+
+    assert (await validate.async_validate(hass)).as_dict() == {
+        "errors": [],
+        "warnings": [],
+        "energy_sources": [
+            {
+                "errors": [
+                    {
+                        "message": "Entity sensor.solar_production has an unexpected unit. Expected kWh or Wh",
+                        "link": None,
+                    }
+                ],
+                "warnings": [],
+            }
+        ],
+        "device_consumption": [],
+    }
+
+
+async def test_validation_battery(hass, mock_energy_manager):
+    """Test validating missing stat for device."""
+    await mock_energy_manager.async_update(
+        {
+            "energy_sources": [
+                {
+                    "type": "battery",
+                    "stat_energy_from": "sensor.battery_import",
+                    "stat_energy_to": "sensor.battery_export",
+                }
+            ]
+        }
+    )
+    hass.states.async_set(
+        "sensor.battery_import", "10.10", {"unit_of_measurement": "beers"}
+    )
+    hass.states.async_set(
```

Comment:
For each test I'd add a source where the validation is expected to succeed for the same test case, to make sure we test both fail and success logic.